src/opencl/mod.rs

```diff
@@ -152,45 +215,78 @@ impl Device {
     }
 }
 
-#[allow(clippy::upper_case_acronyms)]
 #[derive(Debug, Clone, Copy)]
 pub enum GPUSelector {
-    BusId(u32),
+    Uuid(DeviceUuid),
+    PciId(u32),
     Index(usize),
 }
 
 impl GPUSelector {
-    pub fn get_bus_id(&self) -> Option<u32> {
+    pub fn get_uuid(&self) -> Option<DeviceUuid> {
         match self {
-            GPUSelector::BusId(bus_id) => Some(*bus_id),
-            GPUSelector::Index(index) => get_device_bus_id_by_index(*index),
+            GPUSelector::Uuid(uuid) => Some(*uuid),
+            GPUSelector::PciId(pci) => {
+                let mut uuid = None;
+                for dev in Device::all_iter() {
+                    if dev.pci_id == Some(*pci) {
+                        uuid = dev.uuid;
+                        break;
+                    }
+                }
+                uuid
+            }
+            GPUSelector::Index(index) => get_device_uuid_by_index(*index),
+        }
+    }
+
+    pub fn get_pci_id(&self) -> Option<u32> {
+        match self {
+            GPUSelector::Uuid(uuid) => {
+                let mut pci_id = None;
+                for dev in Device::all_iter() {
+                    if dev.uuid == Some(*uuid) {
+                        pci_id = dev.pci_id;
+                        break;
+                    }
+                }
+                pci_id
+            }
+            GPUSelector::PciId(id) => Some(*id),
+            GPUSelector::Index(index) => get_device_pci_id_by_index(*index),
         }
     }
 
     pub fn get_device(&self) -> Option<&'static Device> {
         match self {
-            GPUSelector::BusId(bus_id) => Device::by_bus_id(*bus_id).ok(),
+            GPUSelector::Uuid(uuid) => Device::all_iter().find(|d| d.uuid == Some(*uuid)),
+            GPUSelector::PciId(pci_id) => Device::all_iter().find(|d| d.pci_id == Some(*pci_id)),
             GPUSelector::Index(index) => get_device_by_index(*index),
         }
     }
 
     pub fn get_key(&self) -> String {
         match self {
-            GPUSelector::BusId(id) => format!("BusID: {}", id),
+            GPUSelector::Uuid(uuid) => format!("Uuid: {}", uuid),
+            GPUSelector::PciId(id) => format!("PciId: {}", id),
             GPUSelector::Index(idx) => {
-                if let Some(id) = self.get_bus_id() {
-                    format!("BusID: {}", id)
-                } else {
-                    format!("Index: {}", idx)
-                }
+                format!("Index: {}", idx)
             }
         }
     }
 }
 
-fn get_device_bus_id_by_index(index: usize) -> Option<BusId> {
+fn get_device_uuid_by_index(index: usize) -> Option<DeviceUuid> {
     if let Some(device) = get_device_by_index(index) {
-        device.bus_id
+        device.uuid
+    } else {
+        None
+    }
+}
```

Comment:
I see that you follow the existing coding style (which is good), though I propose making it more idiomatic, while you are touching the code and changing it to:
```rust
get_device_by_index(index).and_then(|device| device.uuid)
````

Reply:
yes. I did not notice the `get_device()` function, using it makes those functions( `get_device_uuid/pci_id_by_index`) unnecessary. Unless I make them public again. what do you think?

Reply:
I don't think they need to be public. In case they are not used in `bellperson` and `neptune` it's fine to make them private.